// purescript-psa/output/Data.Date.UTC/index.js
// Generated by psc version 0.8.5.0
"use strict";
var $foreign = require("./foreign");
var Data_Date = require("../Data.Date");
var Data_Enum = require("../Data.Enum");
var Data_Function = require("../Data.Function");
var Data_Maybe = require("../Data.Maybe");
var Data_Maybe_Unsafe = require("../Data.Maybe.Unsafe");
var Data_Time = require("../Data.Time");
var Prelude = require("../Prelude");

/**
 *  | Gets the UTC year component for a date.
 */
var year = function (d) {
    return $foreign.dateMethod("getUTCFullYear", d);
};

/**
 *  | Get the UTC second-of-minute value for a date.
 */
var secondOfMinute = function (d) {
    return $foreign.dateMethod("getUTCSeconds", d);
};

/**
 *  | Gets the UTC month component for a date.
 */
var month = function (d) {
    return Data_Maybe_Unsafe.fromJust(Data_Enum.toEnum(Data_Date.enumMonth)($foreign.dateMethod("getUTCMonth", d)));
};

/**
 *  | Gets the UTC minute-of-hour value for a date.
 */
var minuteOfHour = function (d) {
    return $foreign.dateMethod("getUTCMinutes", d);
};

/**
 *  | Get the UTC millisecond-of-second value for a date.
 */
var millisecondOfSecond = function (d) {
    return $foreign.dateMethod("getUTCMilliseconds", d);
};

/**
 *  | Gets the UTC hour-of-day value for a date.
 */
var hourOfDay = function (d) {
    return $foreign.dateMethod("getUTCHours", d);
};

/**
 *  | Gets the UTC day-of-week value for a date.
 */
var dayOfWeek = function (d) {
    return Data_Maybe_Unsafe.fromJust(Data_Enum.toEnum(Data_Date.enumDayOfWeek)($foreign.dateMethod("getUTCDay", d)));
};

/**
 *  | Gets the UTC day-of-month value for a date.
 */
var dayOfMonth = function (d) {
    return $foreign.dateMethod("getUTCDate", d);
};

/**
 *  | Attempts to create a `Date` from UTC date and time components. `Nothing`
 *  | is returned if the resulting date is invalid.
 */
var dateTime = function (y) {
    return function (mo) {
        return function (d) {
            return function (h) {
                return function (mi) {
                    return function (s) {
                        return function (ms) {
                            return Data_Date.fromJSDate($foreign.jsDateFromValues(y, Data_Enum.fromEnum(Data_Date.enumMonth)(mo), d, h, mi, s, ms));
                        };
                    };
                };
            };
        };
    };
};

/**
 *  | Attempts to create a `Date` from UTC date components. `Nothing` is
 *  | returned if the resulting date is invalid.
 */
var date = function (y) {
    return function (m) {
        return function (d) {
            return dateTime(y)(m)(d)(0)(0)(0)(0);
        };
    };
};
module.exports = {
    millisecondOfSecond: millisecondOfSecond, 
    secondOfMinute: secondOfMinute, 
    minuteOfHour: minuteOfHour, 
    hourOfDay: hourOfDay, 
    dayOfWeek: dayOfWeek, 
    dayOfMonth: dayOfMonth, 
    month: month, 
    year: year, 
    date: date, 
    dateTime: dateTime
};
