// purescript-psa/output/Node.FS.Stats/index.js
// Generated by psc version 0.8.5.0
"use strict";
var $foreign = require("./foreign");
var Prelude = require("../Prelude");
var Data_Date = require("../Data.Date");
var Data_Function = require("../Data.Function");
var Data_Maybe_Unsafe = require("../Data.Maybe.Unsafe");

/**
 *  | Stats wrapper to provide a usable interface to the underlying properties and methods.
 */
var Stats = (function () {
    function Stats(value0) {
        this.value0 = value0;
    };
    Stats.create = function (value0) {
        return new Stats(value0);
    };
    return Stats;
})();
var statusChangedTime = function (v) {
    return Data_Maybe_Unsafe.fromJust(Data_Date.fromJSDate(v.value0.ctime));
};
var showStats = new Prelude.Show(function (v) {
    return "Stats " + $foreign.showStatsObj(v.value0);
});
var modifiedTime = function (v) {
    return Data_Maybe_Unsafe.fromJust(Data_Date.fromJSDate(v.value0.mtime));
};
var isSymbolicLink = function (v) {
    return $foreign.statsMethod("isSymbolicLink", v.value0);
};
var isSocket = function (v) {
    return $foreign.statsMethod("isSocket", v.value0);
};
var isFile = function (v) {
    return $foreign.statsMethod("isFile", v.value0);
};
var isFIFO = function (v) {
    return $foreign.statsMethod("isFIFO", v.value0);
};
var isDirectory = function (v) {
    return $foreign.statsMethod("isDirectory", v.value0);
};
var isCharacterDevice = function (v) {
    return $foreign.statsMethod("isCharacterDevice", v.value0);
};
var isBlockDevice = function (v) {
    return $foreign.statsMethod("isBlockDevice", v.value0);
};
var accessedTime = function (v) {
    return Data_Maybe_Unsafe.fromJust(Data_Date.fromJSDate(v.value0.atime));
};
module.exports = {
    Stats: Stats, 
    statusChangedTime: statusChangedTime, 
    modifiedTime: modifiedTime, 
    accessedTime: accessedTime, 
    isSymbolicLink: isSymbolicLink, 
    isSocket: isSocket, 
    isFIFO: isFIFO, 
    isCharacterDevice: isCharacterDevice, 
    isBlockDevice: isBlockDevice, 
    isDirectory: isDirectory, 
    isFile: isFile, 
    showStats: showStats
};
