// purescript-psa/output/Data.Date.Locale/index.js
// Generated by psc version 0.8.5.0
"use strict";
var $foreign = require("./foreign");
var Control_Monad_Eff = require("../Control.Monad.Eff");
var Data_Date = require("../Data.Date");
var Data_Enum = require("../Data.Enum");
var Data_Function = require("../Data.Function");
var Data_Maybe = require("../Data.Maybe");
var Data_Maybe_Unsafe = require("../Data.Maybe.Unsafe");
var Data_Time = require("../Data.Time");
var Prelude = require("../Prelude");

/**
 *  | Gets the year component for a date based on the current machine’s locale.
 */
var year = function (d) {
    return $foreign.dateMethod("getFullYear", d);
};

/**
 *  | Format a time as a human-readable string, based on the current machine's
 *  | locale. Example output: "19:45:07", although bear in mind that this
 *  | can vary significantly across platforms.
 */
var toLocaleTimeString = function (d) {
    return $foreign.dateMethod("toLocaleTimeString", d);
};

/**
 *  | Format a date as a human-readable string (including the date and the
 *  | time), based on the current machine's locale. Example output:
 *  | "Fri May 22 2015 19:45:07 GMT+0100 (BST)", although bear in mind that this
 *  | can vary significantly across platforms.
 */
var toLocaleString = function (d) {
    return $foreign.dateMethod("toLocaleString", d);
};

/**
 *  | Format a date as a human-readable string, based on the current machine's
 *  | locale. Example output: "Friday, May 22, 2015", although bear in mind that
 *  | this can vary significantly across platforms.
 */
var toLocaleDateString = function (d) {
    return $foreign.dateMethod("toLocaleDateString", d);
};

/**
 *  | Get the second-of-minute value for a date based on the current machine’s
 *  | locale.
 */
var secondOfMinute = function (d) {
    return $foreign.dateMethod("getSeconds", d);
};

/**
 *  | Gets the month component for a date based on the current machine’s locale.
 */
var month = function (d) {
    return Prelude["<$>"](Control_Monad_Eff.functorEff)(function ($0) {
        return Data_Maybe_Unsafe.fromJust(Data_Enum.toEnum(Data_Date.enumMonth)($0));
    })($foreign.dateMethod("getMonth", d));
};

/**
 *  | Gets the minute-of-hour value for a date based on the current machine’s
 *  | locale.
 */
var minuteOfHour = function (d) {
    return $foreign.dateMethod("getMinutes", d);
};

/**
 *  | Get the millisecond-of-second value for a date based on the current
 *  | machine’s locale.
 */
var millisecondOfSecond = function (d) {
    return $foreign.dateMethod("getMilliseconds", d);
};

/**
 *  | Gets the hour-of-day value for a date based on the current machine’s
 *  | locale.
 */
var hourOfDay = function (d) {
    return $foreign.dateMethod("getHours", d);
};

/**
 *  | Gets the day-of-week value for a date based on the current machine’s
 *  | locale.
 */
var dayOfWeek = function (d) {
    return Prelude["<$>"](Control_Monad_Eff.functorEff)(function ($1) {
        return Data_Maybe_Unsafe.fromJust(Data_Enum.toEnum(Data_Date.enumDayOfWeek)($1));
    })($foreign.dateMethod("getDay", d));
};

/**
 *  | Gets the day-of-month value for a date based on the current machine’s
 *  | locale.
 */
var dayOfMonth = function (d) {
    return $foreign.dateMethod("getDate", d);
};

/**
 *  | Attempts to create a `Date` from date and time components based on the
 *  | current machine’s locale. `Nothing` is returned if the resulting date is
 *  | invalid.
 */
var dateTime = function (y) {
    return function (mo) {
        return function (d) {
            return function (h) {
                return function (mi) {
                    return function (s) {
                        return function (ms) {
                            return Prelude["<$>"](Control_Monad_Eff.functorEff)(Data_Date.fromJSDate)($foreign.jsDateFromValues(y, Data_Enum.fromEnum(Data_Date.enumMonth)(mo), d, h, mi, s, ms));
                        };
                    };
                };
            };
        };
    };
};

/**
 *  | Attempts to create a `Date` from date components based on the current
 *  | machine’s locale. `Nothing` is returned if the resulting date is invalid.
 */
var date = function (y) {
    return function (m) {
        return function (d) {
            return dateTime(y)(m)(d)(0)(0)(0)(0);
        };
    };
};
module.exports = {
    toLocaleDateString: toLocaleDateString, 
    toLocaleTimeString: toLocaleTimeString, 
    toLocaleString: toLocaleString, 
    millisecondOfSecond: millisecondOfSecond, 
    secondOfMinute: secondOfMinute, 
    minuteOfHour: minuteOfHour, 
    hourOfDay: hourOfDay, 
    dayOfWeek: dayOfWeek, 
    dayOfMonth: dayOfMonth, 
    month: month, 
    year: year, 
    date: date, 
    dateTime: dateTime
};
